[file] export MAGIC path to runtime env

themightychris opened the pull request:
`core/file` provides the traditionally-named `/usr/share/misc/magic.mgc` file that some tooling relies on to look for magic bytes in content and determine file formats.

`$MAGIC` seems to be the conventional environment variable for overriding where to find this file

With this change, any plan adding `core/file` to its `pkg_deps` will get `MAGIC` pointing to this file in its runtime env

predominant commented:
@themightychris This appears to be failing the shellcheck due to the `pkg_name`. Can you please adjust this to allow buildkite to pass?

As this is a base plan, I believe this PR should be resubmitted against the base plans refresh branch `refresh/2019q3`. Pinging @smacfarlane for confirmation.

themightychris commented:
@predominant I've corrected the shellcheck issue, rebased against current master, and added some basic tests

smacfarlane commented:
@themightychris sorry for the slow response on this. Because this is a base plan, we'd want to incorporate it into the refresh cycle. I want to move out of an abundance of caution though and take the time to think this change through, given where this software sits in the stack and making updates is an expensive operation to base plans.

I'm trying to reason through the ramifications of this change. Would it be possible to link to a concrete example of where this is needed, either in code or some piece of software that isn't behaving correctly? Having something concrete I can look at helps me unpack these sorts of changes to make sure I'm looking at it from all angles.

Why I'm a little hesitant to accept this change, is that `MAGIC` is a fairly generic variable name, and by setting it in the runtime environment of this package, everything downstream of it will automatically have it set. This may be the desired state in some cases, but could suddenly introduce unexplained behavior to a user.

In addition, doing a quick read of the code, `MAGIC` looks be an escape hatch for users who need to use custom magic files, or files in non-standard locations.  While the magic file provided by Habitat is in a non-standard location, the `file` binary and `libmagic.so` should automatically use the packaged version, and we'd want to reserve `MAGIC` for that escape hatch.  If a piece of software is relying on the magic file, but isn't using `file` or `libmagic.so`, then I would expect it to be that softwares responsibility to set `MAGIC`.  

Please let me know if any of that doesn't track, or if there's a case I'm not considering.

themightychris commented:
@smacfarlane yes those concerns make sense and I agree it's risky. I think there's a good argument that the potential benefit of making more packages "just work" is outweighed by the risk of introducing a new env var to all dependents at this point.

If this were a new package being added I think it would be a good practice to export any conventional env vars for pointing consumers at a provided resource in a nonstandard path, but in this case consuming packages can "opt in" instead by declaring the env var themselves.

PHP's `fileinfo` module is what I added this for.

Would it be helpful if I reworked this to just contribute the tests?

smacfarlane commented:
Interesting

>The libmagic library is bundled with PHP, but includes PHP specific changes. A patch against libmagic named libmagic.patch is maintained and may be found within the PHP fileinfo extensions source. 
https://www.php.net/manual/en/fileinfo.installation.php

That makes me wonder if our builds of PHP are using the bundled version or not.

Tests are always appreciated :heart:  Some documentation around the `MAGIC` variable would probably be a useful addition as well.
